audio/ffmpeg.py

```diff
@@ -0,0 +1,200 @@
+__all__ = ["BlissFFMPEGJob"]
+
+import copy
+import os
+import subprocess
+
+from sisyphus import *
+
+from recipe.i6_core.lib import corpus
+
+
+class BlissFFMPEGJob(Job):
+    """
+    Applies an FFMPEG audio filter to all recordigns of a bliss corpus.
+    This Job is extremely generic, as any valid audio option/filter string will work.
+
+    WARNING:
+        - This Job requires unique file names for the audio files
+        - Changing the duration of the audio files when you have multiple segments per audio,
+          as the segment information will be incorrect afterwards. Please use/write specific
+          jobs in that case (e.g. see corpus.data_augmentation.ChangeCorpusSpeed).
+
+    Typical applications:
+
+    **Changing Audio Format/Encoding**
+
+        - specify in `output_format` what container you want to use. If
+          the filter string is empty (""), ffmepg will automatically use a default encoding option
+
+        - specify specific encoding with `-c:a <codec>`. For a list of available codecs
+          and their options see https://ffmpeg.org/ffmpeg-codecs.html#Audio-Encoders
+
+        - specify a fixed bitrate with `-b:a <bit_rate>`, e.g. `64k`. Variable bitrate options depend on the
+          used encoder, refer to the online documentation in this case
+
+        - specify a sample rate with `-ar <sample_rate>`. FFMPEG will do proper resampling,
+          so the speed of the audio is NOT changed.
+
+
+    **Changing Channel Layout**
+
+        - for detailed informations see https://trac.ffmpeg.org/wiki/AudioChannelManipulation
+
+        - convert to mono `-ac 1`
+
+        - selecting a specific audio channel:
+          `-filter_complex \"[0:a]channelsplit=channel_layout=stereo:channels=FR[right]\" -map \"[right]\"`
+          For a list of channels/layouts use `ffmpeg -layouts`
+
+
+    **Simple Filter Syntax**
+
+    For a list of available filters see: https://ffmpeg.org/ffmpeg-filters.html
+
+    `-af \"<filter_name>=<first_param>=<first_param_value>:<second_param>=<second_param_value>\"`
+
+
+    **Complex Filter Syntax**
+
+    `-filter_complex \"[<input>]<simple_syntax>[<output>];[<input>]<simple_syntax>[<output>];...\"`
+
+    Inputs and outputs can be namend arbitrarily, but the default stream 0 audio can be accessed with [0:a]
+
+    The output stream that should be written into the audio is defined with `-map "[<output_stream>]"`
+
+    IMPORTANT! Do not forget to escape your quotation marks correctly for `-af` or `-filter_complex`
+
+    """
+
+    def __init__(
+        self,
+        corpus_file,
+        ffmpeg_option_string,
+        ffmpeg_binary=None,
```

Comment:
Does it make sense that the ffmpeg binary is part of the hash? That implies that hashes between different systems are not really compatible (e.g. one system has `/usr/bin/ffmpeg` and the other has `/usr/local/bin/ffmpeg` or so). Also, it doesn't tell much (e.g. it doesn't tell what FFmpeg version this is).

One goal of the i6_core recipes is that we can also share data (datasets including preprocessed data such as features) across different systems, and this would break it.

Reply:
I had to do it for myself, because different ffmpeg versions would produce different results for very specific filters and I needed to distinguish this in my pipeline. 
We could actually call "ffmpeg -version" in the hash function and then use this as hash. But then it might be difficult to install exactly the desired version on a system.

Another solution is to write specific jobs that inherit from this one, and allow only for limited parameters, which are then known to be stable across ffmpeg versions.

But no matter the approach, there will be difficulties/inconsistencies.

Reply:
I think in any case it should be made such that things can be shared (also across systems).

Also, as argued, just the path (`/usr/bin/ffmpeg`) doesn't really tell you anything. You update from Ubuntu 16 to Ubuntu 20, path stays the same, but you will get a different FFmpeg.

This problem of different behavior depending on version (and maybe also other things, such as Linux distribution, Linux kernel version, CPU type, whatever) is sth we anyway have. Maybe we should open a separate discussion (separate issue) on how to deal with this in general. I also don't have a good solution. Maybe you can just put some dummy string as argument, sth like `:param str comment:`, and you can write sth like "my custom FFmpeg which behaves different" to explicitly enforce a different hash.

In the usual case, we should assume that we get the same output though, even across different FFmpeg versions.